### DISCUSSION TITLE
Optional 속성의 default value 문제 + React.FC 사용 논의

### DISCUSSION THREAD

**Discussion by hemudi** [Co-Studo/Co-Studo-front#22] [Q&A]
현재 제 프로젝트 환경에서 optinal 속성의 type을 정의하려고 하면 아래와 같은 eslint 에러가 뜨고 있습니다!

![image](https://user-images.githubusercontent.com/34249911/192323117-a9f20876-c0e7-4e6b-8333-79a52ec07a41.png)

[ESLint : Require Default Props](https://github.com/jsx-eslint/eslint-plugin-react/blob/master/docs/rules/require-default-props.md)
해당 에러는 필수 props가 아닌 optional props를 추가하려고 하면 기본 값을 지정해줘야 한다는 에러입니다.
이와 관련하여 질문이 있어 Discussions에 올려봅니다!

Q1. 혹시 이 에러가 제 로컬 환경에서만 뜨고 있는 걸까요? 다들 Optional Prop의 type을 어떻게 정의하고 계신가요?

Q2. 해당 에러의 해결 방법은 공식 문서에서 안내하듯이 해당 컴포넌트의 defaultProps를 사용해 기본값을 넣어주거나, props로 받을 때 default value 를 추가해주는 방식이 있습니다.

```ts
// 1. defaultProps 사용
Component.defaultProps = {
   selectedValue: undefined,
}

// 2. props 에 기본 값 정의
const Component = ({selectedValue = undefined}) => {}
```

그런데 사실 저는 지금까지 프로젝트를 해오면서 해당 에러를 보게 되면 컴포넌트에 defaultProps 를 따로 추가하거나 optional 속성들에 기본 값을 일일이 정의하는게 번거로운 것 같고 코드도 좀 복잡해지는 것 같아서 아래와 같이 그냥 eslint 설정에서 해당 규칙을 off 처리를 해왔습니다. 

```ts
"rules": {
  "react/require-default-props": "off" // or "warn"
}
```

혹시 이 방법들 중 어떤 해결 방법이 낫다고 생각하시는지 궁금합니다!

ㅠ-ㅠ)/ @jindonyy @healtheloper @mina-gwak help!

**Comment by jindonyy** [comment]
저도 안그래도 이 에러가 떴었어여!
근데 제 생각엔 defaul prop이 있는게 맞다고 생각해요. required 가 아니라 해도 그걸 사용하는 로직이 있을테니 해당 로직에서 prop이 없을 때 어떤 걸 사용할건지 명시해줘야한다고 생각해요.
해당 prop이 있는지 없는지 체크하는 로직이라면 default prop이 필요없겠지만 저는 prop으로 숫자들이 넘어오고 해당 숫자들로 계산하는 식이 있어 default prop이 반드시 필요한데요. 해당 옵션을 꺼버리면 제가 실수로 default prop을 빼먹는 일이 생길수도 있으니? (물론 eslint에서 undefined를 어떻게 계산하냐고 난리치긴 합니다. 하하) 그래서 prop이 있는지 없는지 체크하는 로직을 default prop에 undefined나 빈문자열 등을 설정해서 사용하는게 맞지 않을까 싶네요!
근데 궁금한게 저는 2번 방식으로 주로 하는데 2번으로 해도 default prop 넣으라고 에러를 뿜어대서 일단 off 해두고 여쭤보려고 했었어요. 2번으로 해도 잘되나요??

**Reply by hemudi** [reply]
@jindonyy 
저는 type을 optional 로 설정한 순간 해당 prop 이 undefined 로 들어올 수 있다는 걸 허용하는거라고 생각해서 딱히 기본 값을 일괄 적으로 설정해줘야 하나는 의문이 있었어요!
값이 없을 때 대체 할 값이 있는 prop 들만 따로 정의해주면 되지 undefined가 들어와도 상관없는 props까지 전부 undefined로 해주는게 맞나? 싶어서 번거롭다고 생각했습니다! 

그런데 확실히 optional 인 경우에도 도니가 들어준 예처럼 기본 값을 지정해줘야 하는 경우가 있을 수 있으니 그걸 사용하는 부분에서 체크하는 로직을 넣는 것보단 defaultProps로 정의하는게 좀 더 깔끔할 것 같고
이왕 기본 값을 지정하기 시작한다면 undefined나 null이 들어오는 props라 하더라도 일괄적으로 defaultProps로 기본 값을 명시를 해주는 게 좋을 것 같다는 생각도 드네요! ^_^)b

2번 방식의 경우 해당 props를 쓰는 곳에서 전부 다 기본 값을 지정해줘야 해서 빼먹은 부분이 있다면 자꾸 에러가 뜨더라구요...!ㅠ
그래서 개인적으로 에러를 제거하려는 용도라면 2번은 비추인거 같아요!

**Reply by hemudi** [reply]
헉....방금 찾은 글인데 defaultProps가 원래는 클래스형 컴포넌트를 위해 나온 기능이고
앞으로 함수형 컴포넌트에서는 `deprecated` 처리 될 여지가 있다고 하네요ㅠ_ㅠ)

[defaultProps에 대한 짧은 고찰](https://merrily-code.tistory.com/25)
[React defaultProps is dying, who’s the contender?](https://medium.com/@matanbobi/react-defaultprops-is-dying-whos-the-contender-443c19d9e7f1)

이 글을 보니 언젠가 deprecated 될 수도 있는데 써도 괜찮을까 싶은 생각이 드네요ㅠ0ㅠ)

**Comment by healtheloper** [comment]
저도 이전에 비슷한 현상을 봤었고, 함수형에서는 맞지 않는 방법이라해서 단순히 off 해서 썼었는데요, ([stackoverflow](https://stackoverflow.com/questions/63696724/eslint-problem-with-default-props-in-functional-component-typescript-react))
 
컴포넌트의 타입을 React.FC 로 사용하고 Props 정보를 넘겨주면 해당 오류가 나지 않네요, ([stackoverflow](https://stackoverflow.com/a/62610642))
저번에 이야기한 컴포넌트에 타입을 줘야할 이유가 이거인듯?!@

<img width="923" alt="스크린샷 2022-09-27 오전 10 29 35" src="https://user-images.githubusercontent.com/58503584/192410304-d17c0001-daa5-4aa6-beea-aa2ff25cdb51.png">

**Reply by hemudi** [reply]
### 수정 전
![와!!!!!!](https://user-images.githubusercontent.com/34249911/192418107-bc2e7e21-70d7-4b0f-a7ee-b767af33013c.PNG)

### 수정 후
![와웅](https://user-images.githubusercontent.com/34249911/192418476-6d22b8d6-933f-4842-82d1-6268e928a266.PNG)

와!!! ^0^)b
해결됐습니다!!!
ESLint 옵션은 그대로 두고 파크가 말씀해주신 방식으로 통일하는게 좋겠네요!
감사합니다!!!!

**Reply by healtheloper** [reply]
도니가 공유해주신 글 + React.FC 를 쓰면 안된다 ? 에 대해서 좀 더 자세히 알아보고 싶어서 찾아봤는데,

도니가 공유해주신 글에서 출처로 남긴 PR 은 
CRA 의 typescript 기본 template 에 App 의 기본적인 타입을 React.FC 로 지정하는 것을 반대한다는 의견이었어요.

이유는 블로그에서도 간단히 얘기하고는 있지만, React.FC 라는 타입이 best component type 임을 보여주면 안된다는 것이 주 인것 같아요
여러 이유를 이야기하지만, 그 중에서 [children 을 컴포넌트에서 렌더링하거나 처리하는 코드가 없음에도 오류가 안난다는 점](https://yceffort.kr/2022/03/dont-use-react-fc#reactfc%EB%8A%94-%ED%95%AD%EC%83%81-children%EC%9D%84-%EA%B0%80%EC%A7%88%EC%88%98-%EC%9E%88%EB%8B%A4)이었는데 [해당 부분은 v18 로 넘어오면서 수정되었고, children 에 대한 타입을 명시하도록 변경](https://blog.shiren.dev/2022-04-28/) 되었어요.
아래 코드는 v17 에서는 오류가 안나지만, v18 에서는 오류가 납니다.

```tsx
// 오류 안남, codesandbox 는 index.d.ts 가 v17
// https://codesandbox.io/s/heuristic-feistel-mlmmvk?file=/src/App.tsx:0-372

import "./styles.css";

type GreetingProps = {
  name: string;
};

const Greeting: React.FC<GreetingProps> = ({ name }) => {
  return <h1>Hello {name}</h1>;
};

export default function App() {
  return (
    <div className="App">
      <Greeting name="Stefan">
        <span>{"I can set this element but it doesn't do anything"}</span>
      </Greeting>
    </div>
  );
}

```

그럼 이제 React.FC 를 사용하지 말아야 할 이유(?) 라고 이야기 하고 있는 부분들에서

1. 제네릭을 지원하지 않음 -> 컴포넌트에 제네릭을 표현할 상황이 언젠지 상상이 안됨
2. ~~children을 암시적으로 허용합니다. 리턴하는 부분에서 사용하지 않는 경우에도 쓸모없이 전달할 수 있습니다.~~ -> v18 에서 변경됨
3. 컴파운드 컴포넌트 패턴의 타이핑을 혼란스럽게 합니다. -> 이건 동의
4. Component.defaultProps 와 사용하면 이상하다. 혹은 지원하지 않는다. -> default parameter 있는데 굳이..? 

인데, 저는 **2번이 가장 크리티컬한 부분이고, 해결이 되었기 때문에** 일단은 React.FC 로 통일해서 쓰자 라는 의견을 그대로 가져가는게 좋다고 생각합니다!

참고)
[도니가 말한 블로그 글](https://itchallenger.tistory.com/641)
[React.FC를 사용하지 않는 이유 ](https://yceffort.kr/2022/03/dont-use-react-fc#reactfc%EB%8A%94-%ED%95%AD%EC%83%81-children%EC%9D%84-%EA%B0%80%EC%A7%88%EC%88%98-%EC%9E%88%EB%8B%A4)

**Reply by mina-gwak** [reply]
개발하면서 Optional한 파라미터가 생겨서 파크가 말씀하신 대로 적용을 해봤는데, `React`나 `FC`를 import해서 사용하는 경우 동일한 에러가 발생하는 것을 확인할 수 있었습니다. React를 import하라는 인텔리제이의 경고 문구가 떴기 때문에 이런 방식으로 사용한 것이어서 preference에서 관련 설정을 꺼줘도 될 것 같긴 하나, 동일한 타입을 import해서 썼을 때 에러가 생긴다는 건 `React.FC`를 사용하는 게 근본적인 해결책이 아니어서지 않을까 하는 생각이 들었어요. 관련해서 해결 방법을 좀 더 찾아보면 좋을 것 같습니당!

<img width="779" alt="에러" src="https://user-images.githubusercontent.com/62706988/192673947-f374d179-7fc1-4141-866b-09d2620b244d.png" />

**Reply by healtheloper** [reply]
계속 삽질해보면서 내린 결론은

> @types/react 에 index.d.ts 에서 export as namespace React 로 선언함으로써 jQuery 의 $ 처럼 전역으로 쓸 수 있는 UMD global namespace 가 됨

React.FC 로 불러오는 방식 (import 하지 않고) 은 **UMD global 로 가져오기 때문에 ESLint 의 범위로 못 들어오지 않았나** 라는 생각입니다.

ESLint 가 parserOptions 을 tsconfig 를 통해서 읽는데, tsconfig 는 src 와 webpack 만을 include 하고있기에 UMD global 로 사용되고 있는 react namespace 를 해석하지 못한거(?) 라고 생각했어요, 대신 import 하는 순간 부터는 TypeScript 의 include 영역 내부에 있는 것이니 인식이 되는 거구요

어디까지나 추측이지만 일단은 이게 가장 이치가 맞다고 생각하렵니다.
그러면 어떻게 보면 편법처럼 eslint 룰을 무시(?) 하고 있는 거고, 사실상 function component 를 쓰는 상황에서 defaultProps 는 dying 이라고 하니 **off 하는 것** 에 의견내봅니다~!

+) 저희와 비슷한 상황을 [eslint issue](https://github.com/jsx-eslint/eslint-plugin-react/issues/3336) 로 등록해놓은 것도 있네요

참고)
tsconfig 의 [allowUMDGlobalAccess](https://evan-moon.github.io/2021/08/22/tsconfig-compiler-options-modules/#allowumdglobalaccess)
[React defaultProps is dying, who’s the contender?](https://medium.com/@matanbobi/react-defaultprops-is-dying-whos-the-contender-443c19d9e7f1)

**Reply by hemudi** [reply]
헉...eslint 의 범위를 이렇게 벗어나는 경우가 있을 수 있군요.
일리 있는 추측이신 것 같아요! 분석해주셔서 감사합니다!
저도 이러한 이유라면 off에 한 표 던져봅니다!

그리고 만약 해당 옵션을 off하면 "[Headless 컴포넌트의 하위 컴포넌트 타입 지정 방식](https://github.com/Co-Studo/Co-Studo-front/discussions/25)" 이 글에 대한 논의도 그냥 원래의 방식대로 유지하는걸로 결론지어도 괜찮을 것 같기도 하네요!

**Reply by mina-gwak** [reply]
저도 동의합니다!
찾아봐도 정확히 나오진 않지만 파크가 추측하신 게 맞지 않을까 싶네요.
@hemudi 그런데 [하위 컴포넌트 타입에 대한 문제](https://github.com/Co-Studo/Co-Studo-front/discussions/25)는 ESLint 옵션을 off하는 것과는 별개의 문제가 아닌가 싶은데 혹시 제가 이해를 잘못한 걸까요..? 옵션을 off해도 해당 에러는 그대로 남아있는 것 같아서요..!

**Reply by jindonyy** [reply]
첨부해주신 eslint issue 글을 보니 추측이 맞는거 같아요..!
@hemudi 가 말하는 원래의 방식이 뭘 말하는걸까요??

1.  [Object.assign](https://github.com/Co-Studo/Co-Studo-front/discussions/25#discussion-4424912) 방식
-> React.FC를 쓰려면 이걸 쓰는게 맞는거 같아요.
2. React.FC를 안쓰던 방식
-> defualt prop 옵션을 off했으니 근본적으로 React.FC를 안써도 되지 않을까?

근데 React.FC를 써야하는 이유를 찾아보니
1. propTypes, contextTypes, defaultProps, displayName이 자동완성 된다.
FC 타입의 내부 코드
```ts
type FC<P = {}> = FunctionComponent<P>

interface FunctionComponent<P = {}> {
  (props: PropsWithChildren<P>, context?: any): ReactElement<any, any> | null
  propTypes?: WeakValidationMap<P> | undefined
  contextTypes?: ValidationMap<any> | undefined
  defaultProps?: Partial<P> | undefined
  displayName?: string | undefined
}
```
2. <s>children이 props 에 기본적으로 children 이 들어가있다. (장점이자 단점)</s> -> v.18에서 개선

이 두 가지던데 2번은 바뀌었으니 1번의 장점을 위해 사용하는건데 굳이 써야하나 싶은 부분이 있어요.
여러 글을 읽어보다 [SFC는 FC가 되었다. 또 훗날 FC가 무엇으로 바뀔 수 있을지도 모를일이다.](https://yceffort.kr/2022/03/dont-use-react-fc#stateless-function-component%EC%9D%98-%EA%B3%BC%EA%B1%B0) 라는 의견에 동의하기 떄문에 1번의 장점떄문이라면 꼭 사용해야할까 싶네요!

**Reply by hemudi** [reply]
앗 제가 생각을 잘못했네요!!!
오류가 나는 원인이 달랐는데 순간 착각했어요!!
죄송합니다 ^//^)>
저도 Object.assign 방식에 동의해요ㅎㅎ!!

+) 제가 말한 원래의 방식은 하위 컴포넌트에 따로 타입을 지정히지 않고 
Radio.Title = Title 로 그냥 하위 컴포넌트를 합성하는 방식을 말하는 거였어요!
ㅎㅎ제이미 말씀 처럼 오류가 나는 원인이 다른데 연속해서 나왔던 이슈라 순간 착각해서 그대로 해도 되나 싶었었어요!

**Reply by hemudi** [reply]
@Dony 그럼 React.FC에 대해서도 다시 새 글을 파서 의견을 모아볼까요?

**Reply by healtheloper** [reply]
넹 poll 로 한번 해보시쥬 ㅋㅋ
